esp-hal/src/dma/mod.rs

```diff
@@ -732,59 +730,68 @@ where
     }
 
     fn available(&mut self) -> usize {
-        if self.last_seen_handled_descriptor_ptr.is_null() {
-            self.last_seen_handled_descriptor_ptr = self.descriptors.as_mut_ptr();
-            return 0;
-        }
-
-        if self.available != 0 {
-            return self.available;
-        }
-
-        let descr_address = self.last_seen_handled_descriptor_ptr;
-        let mut dw0 = unsafe { descr_address.read_volatile() };
+        let last_in_dscr_ptr = self.rx_impl.last_in_dscr_address() as *mut DmaDescriptor;
+        let last_in_descr = unsafe { last_in_dscr_ptr.read_volatile() };
 
-        if dw0.owner() == Owner::Cpu && !dw0.is_empty() {
-            let descriptor_buffer = dw0.buffer;
-            let next_descriptor = dw0.next;
-
-            self.read_buffer_start = descriptor_buffer;
-            self.available = dw0.len();
-
-            dw0.set_owner(Owner::Dma);
-            dw0.set_length(0);
-            dw0.set_suc_eof(false);
+        if !last_in_descr.is_empty() && last_in_descr.owner() == Owner::Cpu {
+            let len = last_in_descr.len();
 
-            unsafe {
-                descr_address.write_volatile(dw0);
+            if self.last_seen_handled_descriptor_ptr.is_null() {
+                self.read_descriptor_ptr = last_in_dscr_ptr;
+                self.available = len;
+            } else if self.last_seen_handled_descriptor_ptr != last_in_dscr_ptr {
+                self.available += len;
```

Comment:
This code is assuming that `last_in_dscr_ptr` is equal to `self.last_seen_handled_descriptor_ptr.next`.
If the user takes a while to call `available()` again after calling it once, there may be multiple descriptors that now available.
So this here should be a while loop looking at all the descriptors between `self.last_seen_handled_descriptor_ptr` and `last_in_dscr_ptr`.

Reply:
true - we already do something similar for TX 👍